treewide: clean up 'meta = with' pattern

booxter opened the pull request:
To reproduce (not tested, just the gist):

```
$ NIXPKGS=~/src/nixpkgs
$ gh repo clone technowledgy/refactor-tractor
$ cd refactor-tractor
$ gh pr checkout 15
$ devenv shell
(devenv) remove-meta-with-lib $NIXPKGS --write
$ cd $NIXPKGS && git commit -a -m "treewide: clean up 'meta = with' pattern"
```

A machine with lots of cores is recommended because nixpkgs-aware variable analysis demands significant compute power to plow through all .nix files in-tree in reasonable time.

<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [x] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

booxter commented on nixos/release.nix:
```diff
@@ -121,7 +121,7 @@ let
     // {
       meta = {
         description = "NixOS system tarball for ${system} - ${stdenv.hostPlatform.linux-kernel.name}";
-        maintainers = map (x: lib.maintainers.${x}) maintainers;
```
@wolfgangwalther we'll probably need to tighten the pattern to match on `meta = with lib; ` only. FYI

Another example of a problematic pattern would be:

```
{
pkgs,
... # no lib argument!
}:
... {
  meta = with pkgs.lib; {
  };
}
```

I've seen it a few times in the tree.

booxter replied:
I will spend some time on the refactoring rules this week. It's on my radar.

booxter commented:
Status of this PR:

- base `refactor-tractor` infrastructure, with auto-fix enabled: https://github.com/technowledgy/refactor-tractor/pull/14 (waiting)
- `nixd` support for nixpkgs-based variable resolution: https://github.com/nix-community/nixd/pull/719 (need review);
- once nixd PR is merged and released, we'll need to pick it up in refactor-tractor: https://github.com/technowledgy/refactor-tractor/pull/15/commits/958fb93d770152a3775d34669147bb01d8f91c58
- nixf-diagnose is now enabled treewide. #438559 

@wolfgangwalther **apart** from nixd and refactor-tractor parts, what else do we miss? E.g. do we still need a verifier for transformations not affecting calculated `meta` attributes? How are backports going to be handled?

booxter commented:
I will write a meta checker later this week. Thanks for spotting these issues.

booxter commented:
Using https://github.com/booxter/nixpkgs-meta-checker on this PR:

```
$ ./diff_all_meta.sh pre post
== Diffing x86_64-linux ==
* sunvox changed
    - platforms: ["aarch64-darwin", "aarch64-linux", "armv7l-linux", "i686-linux", "x86_64-darwin", "x86_64-linux"]
    + platforms: ["aarch", "aarch64", "all", "arm", "armv7", "bigEndian", "cygwin", "darwin", "embedded", "freebsd", "genode", "gnu", "i686", "illumos", "js", "linux", "littleEndian", "loongarch64", "m68k", "microblaze", "mips", "mmix", "netbsd", "none", "openbsd", "or1k", "power", "redox", "riscv", "riscv32", "riscv64", "rx", "s390", "s390x", "unix", "vc4", "wasi", "windows", "x86", "x86_64"]
    - unsupported: false
    + unsupported: true

== Diffing x86_64-darwin ==
* sunvox changed
    - platforms: ["aarch64-darwin", "aarch64-linux", "armv7l-linux", "i686-linux", "x86_64-darwin", "x86_64-linux"]
    + platforms: ["aarch", "aarch64", "all", "arm", "armv7", "bigEndian", "cygwin", "darwin", "embedded", "freebsd", "genode", "gnu", "i686", "illumos", "js", "linux", "littleEndian", "loongarch64", "m68k", "microblaze", "mips", "mmix", "netbsd", "none", "openbsd", "or1k", "power", "redox", "riscv", "riscv32", "riscv64", "rx", "s390", "s390x", "unix", "vc4", "wasi", "windows", "x86", "x86_64"]
    - unsupported: false
    + unsupported: true

== Diffing aarch64-linux ==
* sunvox changed
    - platforms: ["aarch64-darwin", "aarch64-linux", "armv7l-linux", "i686-linux", "x86_64-darwin", "x86_64-linux"]
    + platforms: ["aarch", "aarch64", "all", "arm", "armv7", "bigEndian", "cygwin", "darwin", "embedded", "freebsd", "genode", "gnu", "i686", "illumos", "js", "linux", "littleEndian", "loongarch64", "m68k", "microblaze", "mips", "mmix", "netbsd", "none", "openbsd", "or1k", "power", "redox", "riscv", "riscv32", "riscv64", "rx", "s390", "s390x", "unix", "vc4", "wasi", "windows", "x86", "x86_64"]
    - unsupported: false
    + unsupported: true

== Diffing aarch64-darwin ==
* sunvox changed
    - platforms: ["aarch64-darwin", "aarch64-linux", "armv7l-linux", "i686-linux", "x86_64-darwin", "x86_64-linux"]
    + platforms: ["aarch", "aarch64", "all", "arm", "armv7", "bigEndian", "cygwin", "darwin", "embedded", "freebsd", "genode", "gnu", "i686", "illumos", "js", "linux", "littleEndian", "loongarch64", "m68k", "microblaze", "mips", "mmix", "netbsd", "none", "openbsd", "or1k", "power", "redox", "riscv", "riscv32", "riscv64", "rx", "s390", "s390x", "unix", "vc4", "wasi", "windows", "x86", "x86_64"]
    - unsupported: false
    + unsupported: true
```

So `sunvox` is the only affected package in the whole tree.

---

I am not sure how to make `ast-grep` behave here. We would need to make it NOT apply the rule if it's inside a `let` that defines the target symbol.

I tried the following but it fails for some reason:

```diff
diff --git a/rules/prefix-meta-lib.yml b/rules/prefix-meta-lib.yml
index e8569b7..ef5cbe9 100644
--- a/rules/prefix-meta-lib.yml
+++ b/rules/prefix-meta-lib.yml
@@ -4,7 +4,8 @@ language: nix
 rule:
   pattern: maintainers
   kind: identifier
-  inside:
+  all:
+  - inside:
       # variable_expression is a non-prefixed occurence of "maintainers"
       kind: variable_expression
       inside:
@@ -26,6 +27,14 @@ rule:
               has:
                 field: environment
                 regex: ^lib$
+  - not:
+      inside:
+        kind: let_expression
+        has:
+          kind: binding
+          has:
+            field: attrpath
+            regex: ^maintainers$
 
 fix: lib.maintainers
 
diff --git a/tests/prefix-meta-lib.yml b/tests/prefix-meta-lib.yml
index 6e886b3..7d45130 100644
--- a/tests/prefix-meta-lib.yml
+++ b/tests/prefix-meta-lib.yml
@@ -26,6 +26,15 @@ valid:
       maintainers = with maintainers; [];
     };
   }
+# Should not match with `let` binding
+- |
+  let
+    maintainers = [ "alice" "bob" ];
+  in {
+    meta = with lib; {
+      maintainers = maintainers;
+    };
+  }
 invalid:
 - |
   {
```

If you have ideas how to bend `ast-grep`, let me know. @wolfgangwalther

booxter commented:
Generated using `80c42444a7613782290df3ab5d2e346805620071` commit from: https://github.com/technowledgy/refactor-tractor/pull/15/commits

Changes:
- `hydraPlatforms` and `badPlatforms` are now also handled;
- variables bound inside `let`s are now considered, avoiding spurious overrides when their names match our heuristic patterns like `platforms`;
- all files with `meta = with lib;` are now processed, even when no `ast-grep` modifications happened;
- `nixpkgs` used for variable resolution is now pulled from the same directory as the target for modification (not from system `<nixpkgs>`).

```
$ git grep 'meta = with lib;' | grep '.nix:' | wc -l
90

$ ./diff_all_meta.sh pre post
== Diffing x86_64-linux ==

== Diffing x86_64-darwin ==

== Diffing aarch64-linux ==

== Diffing aarch64-darwin ==
```

No meta changes detected.

booxter commented on nixos/modules/services/desktops/geoclue2.nix:
```diff
@@ -373,6 +373,6 @@ in
   };
 
   meta = with lib; {
```
This is addressed now too.

philiptaron commented:
@wolfgangwalther, are we getting close to merge? I think this would be nice to land.

booxter committed:
treewide: clean up 'meta = with' pattern

This commit was created by a combination of scripts and tools:
- an ast-grep script to prefix things in meta with `lib.`,
- a modified nixf-diagnose / nixf combination to remove unused `with
lib;`, and
- regular nixfmt.

Co-authored-by: Wolfgang Walther <walther@technowledgy.de>
